### PR TITLE
Add promise C API

### PR DESCRIPTION
**Pull request by jiangzidong**
related issue: 1794

* add api: 
```
bool jerry_value_is_promise (const jerry_value_t value);
jerry_value_t jerry_create_promise (void);
jerry_value_t jerry_resolve_promise (jerry_value_t promise, jerry_value_t argument);
jerry_value_t jerry_reject_promise (jerry_value_t promise, jerry_value_t reason);
```

* unittest support es2015-subset profile

* doc will be added later when we reach the consensus

### CODE REVIEW

File: jerry-core/jerry.c [hunk 1]
```diff
@@ -553,6 +554,26 @@ jerry_value_is_object (const jerry_value_t value) /**< api value */
...
+ * Check if the specified value is promise.
+ *
+ * @return true  - if the specified value is promise,
+ *         false - otherwise
+ */
+bool
+jerry_value_is_promise (const jerry_value_t value) /**< api value */
+{
+  jerry_assert_api_available ();
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
```
Comment (zherczeg) [inline jerry-core/jerry.c]: newline before ifdef

File: tests/unit-core/test-promise.c [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+static int count_in_assert = 0;
+static jerry_value_t my_promise1;
+static jerry_value_t my_promise2;
+const jerry_char_t s1[] = "resolved";
+const jerry_char_t s2[] = "rejected";
+
+static jerry_value_t
+create_promise1_handler (const jerry_value_t func_obj_val __attribute__((unused)), /**< function object */
+                        const jerry_value_t this_val __attribute__((unused)), /**< this value */
```
Comment (zherczeg) [inline tests/unit-core/test-promise.c]: One more space needed.

File: tests/unit-core/test-promise.c [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+{
+  jerry_value_t ret =  jerry_create_promise ();
+  my_promise1 = jerry_acquire_value (ret);
+
+  return ret;
+} /* create_promise1_handler */
+
+static jerry_value_t
+create_promise2_handler (const jerry_value_t func_obj_val __attribute__((unused)), /**< function object */
+                        const jerry_value_t this_val __attribute__((unused)), /**< this value */
```
Comment (zherczeg) [inline tests/unit-core/test-promise.c]: One more space needed.

File: jerry-core/jerry.c (hunk 1 above)
Comment (zherczeg) [inline jerry-core/jerry.c]: Please add newlines before other ifdefs in this file.

File: jerry-core/jerry.c [hunk 4]
```diff
@@ -1019,6 +1040,26 @@ jerry_create_object (void)
...
+ * @return value of the created object
+ */
+jerry_value_t
+jerry_create_promise (void)
+{
+  jerry_assert_api_available ();
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+  return ecma_op_create_promise_object (ecma_make_simple_value (ECMA_SIMPLE_VALUE_EMPTY), ECMA_PROMISE_EXECUTOR_EMPTY);
+#else /* CONFIG_DISABLE_ES2015_PROMISE_BUILTIN */
+  return ecma_raise_syntax_error (ECMA_ERR_MSG ("Promise profile has been disabled."));
```
Comment (martijnthe) [inline jerry-core/jerry.c]: Maybe better to exclude the implementation so you get a linker error?
I don't know... returning an error adds code space.

File: jerry-core/jerry.c [hunk 5]
```diff
@@ -2121,6 +2162,85 @@ jerry_foreach_object_property (const jerry_value_t obj_val, /**< object value */
...
 /**
+ * Common part for jerry_resolve_promise and jerry_reject_promise.
+ *
+ * @return undefined value - if success
+ *         value marked with error flag - otherwise
+ */
+static jerry_value_t
+jerry_resolve_or_reject_promise (jerry_value_t promise,
+                                 jerry_value_t argument,
+                                 bool is_resolve)
```
Comment (martijnthe) [inline jerry-core/jerry.c]: I think it makes sense to expose something like this in the public API.

I think people will be writing code like in [this example] a lot, so you naturally have a `bool` to make the decision whether `resolve` or `reject` should be called.

In an [earlier sketch], I used an `enum`, I think that has a bit better readability vs `bool`, but I'm not feeling strongly about it.

[this example]: https://github.com/jerryscript-project/jerryscript/issues/1794#issuecomment-298308945
[earlier sketch]: https://github.com/jerryscript-project/jerryscript/issues/1794#issuecomment-298319724

**Comment by martijnthe** [comment]
@jprestwo any feedback on this API?

File: jerry-core/jerry.c (hunk 5 above)
Comment (jiangzidong) [inline jerry-core/jerry.c]: My thought is the C api should be similar with js api, so it will be easy to use. In JS world, usually the code is like
``` javascript
var p = new Promise(function(resolve, reject)
{
  ....
  if (some condition)
  {
    var reason = .....//prepare the reason
     reject (reason)
  }
  
  var argument = .... //prepare the argument
   resolve(argument)
})
```

in your example, 
``` c
static void my_get_batt_info_cb (const batt_info_t *info, void *cb_data)
{
        jerry_value_t promise = (jerry_value_t)(uintptr_t) cb_data; 
        
        // do stuff with *info and assign result + should_call_resolve:
        jerry_value_t result = ...;
        bool should_call_resolve = ...;
        
        jerry_call_resolve_reject(promise, result, should_call_resolve);
        jerry_value_release(promise);
}
```
there should be some if-else or branch code in `result = ` and `should_call_resolve =` , e.g.
``` c
if (some condition)
 { result = some_error_reason; should_call_resolve = false; }
else
{result = some_argument; should_call_resolve = true;}
```
why just directly call the resolve/reject in the branch?

Comment (martijnthe) [inline jerry-core/jerry.c]: > why just directly call the resolve/reject in the branch?

Code space savings. You'd have 2x function calls vs 1x.
Often you already have the `bool should_call_resolve` one way or another, for example in your snippet, there is `some_condition`. In this case, `some_condition == !should_call_resolve`.

Comment (jiangzidong) [inline jerry-core/jerry.c]: Ok, it makes sense :)

File: docs/02.API-REFERENCE.md [hunk 6]
```diff
@@ -1980,6 +2021,139 @@ jerry_value_to_string (const jerry_value_t value);
...
+Get the resolve function of the promise object.
+
+*Note*: Returned value must be freed with [jerry_release_value](#jerry_release_value) when it
+is no longer needed.
+
+**Prototype**
+
+```c
+jerry_value_t
+jerry_promise_get_resolve (jerry_value_t promise)
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: What I don't understand is that we have one function to resolve/reject, but two functions for getting functions. I think it would be better to have a boolean parameter here as well.

Comment (martijnthe) [inline docs/02.API-REFERENCE.md]: My gut says that if your project needs access to the functions, you probably are going to need both.

Would it make sense to have:

```
bool jerry_promise_get_resolve (
     jerry_value_t promise, jerry_value_t *resolve_out_p, jerry_value_t *reject_out_p)
```

Comment (martijnthe) [inline docs/02.API-REFERENCE.md]: Hmm I wonder if we're adding the resolve/reject getter prematurely...
Maybe just remove it and get on with life until someone comes along that has a clear use for it?

Comment (jiangzidong) [inline docs/02.API-REFERENCE.md]: agree

File: docs/02.API-REFERENCE.md [hunk 7]
```diff
@@ -1980,6 +2021,64 @@ jerry_value_to_string (const jerry_value_t value);
 - [jerry_value_to_primitive](#jerry_value_to_primitive)
 
 
+# Functions for promise value
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: Functions for promise objects.

File: docs/02.API-REFERENCE.md [hunk 8]
```diff
@@ -1980,6 +2021,64 @@ jerry_value_to_string (const jerry_value_t value);
...
+
+**Example**
+
+```c
+{
+  jerry_value_t promise = ... // acquire/create a promise object.
+
+  ...
+
+  bool is_resolve = ... // whether the promise should be resolve or reject
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: resolved or rejected

File: docs/02.API-REFERENCE.md [hunk 9]
```diff
@@ -2425,6 +2524,41 @@ jerry_create_object (void);
 - [jerry_release_value](#jerry_release_value)
 
 
+## jerry_create_promise
+
+**Summary**
+
+Create an empty Promise object which can be resolve/reject later
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: resolved/rejected

File: docs/02.API-REFERENCE.md [hunk 10]
```diff
@@ -2425,6 +2524,41 @@ jerry_create_object (void);
...
+*Note*: This API depends on the ES2015-subset profile.
+
+**Prototype**
+
+```c
+jerry_value_t
+jerry_create_promise (void)
+```
+
+- return value - value of the created promise
```
Comment (zherczeg) [inline docs/02.API-REFERENCE.md]: newly created promise

File: jerry-core/ecma/operations/ecma-promise-object.h [hunk 11]
```diff
@@ -38,6 +38,16 @@ typedef enum
 } ecma_promise_state_t;
 
 /**
+ * Indicates the type of the executor in promise construct.
+ */
+typedef enum
+{
+  ECMA_PROMISE_EXECUTOR_FUNCTION, /**< the executor is a function, it is for the usual constructor */
+  ECMA_PROMISE_EXECUTOR_OBJECT, /**< the executor is a object, it is for the `then` routine */
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-promise-object.h]: the executor is an object

File: jerry-core/jerry.c [hunk 12]
```diff
@@ -2175,6 +2176,84 @@ jerry_set_vm_exec_stop_callback (jerry_vm_exec_stop_callback_t stop_cb, /**< per
...
+jerry_value_is_promise (const jerry_value_t value) /**< api value */
+{
+  jerry_assert_api_available ();
+
+  return (ecma_is_value_object (value)
+          && ecma_is_promise (ecma_get_object_from_value (value)));
+} /* jerry_value_is_promise */
+
+
+/**
```
Comment (zherczeg) [inline jerry-core/jerry.c]: Only one newline.

File: jerry-core/jerry.c [hunk 13]
```diff
@@ -2175,6 +2176,83 @@ jerry_set_vm_exec_stop_callback (jerry_vm_exec_stop_callback_t stop_cb, /**< per
...
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+
+/**
+ * Check if the specified value is promise.
+ *
+ * @return true  - if the specified value is promise,
+ *         false - otherwise
+ */
+bool
+jerry_value_is_promise (const jerry_value_t value) /**< api value */
```
Comment (LaszloLango) [inline jerry-core/jerry.c]: Please use the same order of functions as 'jerryscript.h'.

File: tests/unit-core/test-promise.c [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+                           );
+
+static int count_in_assert = 0;
+static jerry_value_t my_promise1;
+static jerry_value_t my_promise2;
+const jerry_char_t s1[] = "resolved";
+const jerry_char_t s2[] = "rejected";
+
+static jerry_value_t
+create_promise1_handler (const jerry_value_t func_obj_val __attribute__((unused)), /**< function object */
```
Comment (LaszloLango) [inline tests/unit-core/test-promise.c]: `JERRY_UNUSED` must be available here since test-common.h includes jrt.h
```c
static jerry_value_t
create_promise1_handler (const jerry_value_t func_obj_val , /**< function object */
                         const jerry_value_t this_val, /**< this value */
                         const jerry_value_t args_p[], /**< arguments list */
                         const jerry_length_t args_cnt) /**< arguments length */
{
JERRY_UNUSED (func_obj_val);
JERRY_UNUSED (this_val);
JERRY_UNUSED (args_p);
JERRY_UNUSED (args_cnt);

...
```

File: tests/unit-core/test-promise.c [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+                         const jerry_length_t args_cnt __attribute__((unused))) /**< arguments length */
+{
+  jerry_value_t ret =  jerry_create_promise ();
+  my_promise1 = jerry_acquire_value (ret);
+
+  return ret;
+} /* create_promise1_handler */
+
+static jerry_value_t
+create_promise2_handler (const jerry_value_t func_obj_val __attribute__((unused)), /**< function object */
```
Comment (LaszloLango) [inline tests/unit-core/test-promise.c]: ditto

File: tests/unit-core/test-promise.c [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+                         const jerry_length_t args_cnt __attribute__((unused))) /**< arguments length */
+{
+  jerry_value_t ret =  jerry_create_promise ();
+  my_promise2 = jerry_acquire_value (ret);
+
+  return ret;
+} /* create_promise2_handler */
+
+static jerry_value_t
+assert_handler (const jerry_value_t func_obj_val __attribute__((unused)), /**< function object */
```
Comment (LaszloLango) [inline tests/unit-core/test-promise.c]: ditto

File: tests/unit-core/CMakeLists.txt [hunk 17]
```diff
@@ -19,8 +19,8 @@ if (NOT IS_ABSOLUTE ${FEATURE_PROFILE})
   set(FEATURE_PROFILE "${CMAKE_SOURCE_DIR}/jerry-core/profiles/${FEATURE_PROFILE}.profile")
 endif()
 
-if(NOT ${FEATURE_PROFILE} STREQUAL "${CMAKE_SOURCE_DIR}/jerry-core/profiles/es5.1.profile")
-  message(FATAL_ERROR "FEATURE_PROFILE='${FEATURE_PROFILE}' isn't supported with UNITTESTS=ON")
+if(${FEATURE_PROFILE} STREQUAL "${CMAKE_SOURCE_DIR}/jerry-core/profiles/minimal.profile")
+  message(FATAL_ERROR "minimal profile isn't supported in unit-core")
```
Comment (LaszloLango) [inline tests/unit-core/CMakeLists.txt]: Is this change related to this this PR? It looks to me if promise is not available than the test-promise returns zero, which is good. Am I missing something?

Comment (jiangzidong) [inline tests/unit-core/CMakeLists.txt]: Before this change, I cannot `./tools/build.py --unittest --profile=es2015-subset`, because it's profile is not es5.1.

Comment (jiangzidong) [inline tests/unit-core/CMakeLists.txt]: And also I changed the run-test.py, now es2015-subset profile is passed to unittest

```
# Test options for unittests
JERRY_UNITTESTS_OPTIONS = [
    Options('unittests',
            ['--unittests', '--error-messages=on', '--snapshot-save=on', '--snapshot-exec=on', '--vm-exec-stop=on', '--profile=es2015-subset']),
    Options('unittests-debug',
            ['--unittests', '--debug', '--error-messages=on', '--snapshot-save=on', '--snapshot-exec=on', '--vm-exec-stop=on', '--profile=es2015-subset'])
]
```

Comment (LaszloLango) [inline tests/unit-core/CMakeLists.txt]: I see. Thanks for the explanation.

**Comment by martijnthe** [comment]
💯
